### ISSUE TITLE
P4HIR: an experimental MLIR dialect that models P4C IR

### ISSUE THREAD

**Issue by qobilidop** [p4lang/project-ideas#20]
## Update

We are now incubating this idea in https://github.com/p4lang/p4mlir-incubator. @asl is leading the design and development. Modeling the existing P4C IR as it is today is no longer the plan. Check out https://github.com/p4lang/p4mlir-incubator/wiki/IR-Design-Notes for @asl's current design!

For anyone interested in participating and contributing, you are welcome in our incubator repo!

## Original Proposal

This project idea is created to invite further discussion of my 2024 P4 Workshop talk:

- P4HIR: Towards Bridging P4C with MLIR ([slides](https://p4.org/wp-content/uploads/2024/11/204-P4-Workshop-P4HIR_-Towards-Bridging-P4C-with-MLIR-P4-Workshop-2024.pdf), [video](https://youtu.be/0qov53z_ExM)).

P4HIR (P4 High-level IR) is an experimental [MLIR](https://mlir.llvm.org/) [dialect](https://mlir.llvm.org/docs/LangRef/#dialects) that aims to model the existing [P4C IR](https://github.com/p4lang/p4c/blob/main/docs/IR.md) as closely as possible, so it could serve as a straightforward bridge between the existing P4C infra and a potential MLIR ecosystem involving P4.

More detailed design of P4HIR will be figured out during the implementation.

**Comment by fruffy** [comment]
I have also been thinking of starting something like this. Partially because I wanted to learn more about MLIR-style compilers. Is the work planned to be open-source? I would consider contributing it as an extension first (and with that a separate repository), since LLVM is such a large dependency. 

It might be a good idea to have this back end produce code that is executable by a target, e.g., BMv2. Not sure if that is also planned.

**Comment by qobilidop** [comment]
Yes, open-source is the plan! I think having a separate repo is a good idea. I've been working in a local P4C fork in a subdirectory. But it shall be straightforward to move to a separate repo.

**Comment by qobilidop** [comment]
> It might be a good idea to have this back end produce code that is executable by a target, e.g., BMv2. Not sure if that is also planned.

Agree with this! My thought is that the BMv2 JSON format could be defined as another MLIR dialect, and the BMv2 backend could be ported to be transformations from P4HIR dialect to the BMv2 dialect, with a relatively trivial final codegen from BMv2 dialect to BMv2 JSON.

This might be a great idea for a standalone follow-up work, once we have a relatively mature P4HIR that covers enough P4C IR to be useful.

**Comment by qobilidop** [comment]
After talking to @asl today, I learned that there is an existing work that basically does this already: https://github.com/barinsim/p4c-mlir

Here's my summary of what this project does:

- Defined a `P4` dialect. See https://github.com/barinsim/p4c-mlir/blob/main/extensions/mlir/include/P4Ops.td for an overview of the operations defined. It seems to have captured most P4 primitives.
- Implemented the translation from P4C IR to the `P4` dialect in https://github.com/barinsim/p4c-mlir/tree/main/extensions/mlir/p4c-mlir-translate. Also see https://github.com/barinsim/p4c-mlir/tree/main/extensions/mlir/test/translate for examples of this translation.
- Implemented the lowering from `P4` dialect to `LLVM` dialect in https://github.com/barinsim/p4c-mlir/tree/main/extensions/mlir/p4dialect-llvm-translate. This part seems to be WIP.

I feel this could be a pretty good starting point for further development. But as @asl suggested, maybe we should hold one step back and present a detailed design for review before diving into the implementation.

@barinsim If you are still working on this and are interested in upstreaming these work, maybe we could have some discussions?

**Comment by qobilidop** [comment]
Closing this as we now have https://github.com/p4lang/p4mlir-incubator.
